ui/dialog/MintDialog.tsx

```diff
@@ -36,20 +43,90 @@ const DialogFooterButton = ({ children, ...props }: Omit<ButtonProps, 'ref'>) =>
+
+
+  useEffect(() => {
+    ( async () => {
+    if ( coldwallets && whitelist ) {
+      console.log("cold wallets:", coldwallets)
+      const ethersProvider = new ethers.providers.Web3Provider(provider);
+      const pfpContract = new ethers.Contract(pfpAddress, pfpAbi, ethersProvider);
+      let mintAmount = 0;
+      let mintWallet = "";
+      for(let i=0; i<coldwallets.length; i++) {
+        let coldWallet = coldwallets[i];
+        if(whitelist[coldWallet as keyof typeof whitelist]) {
+          let allowance = whitelist[coldWallet as keyof typeof whitelist];
+          let mb = await pfpContract.balanceOf(coldWallet);
+          let balance = parseInt(BigNumber.from(mb).toString())
+          if(mintAmount <= (allowance-balance)) {
+            mintAmount = allowance-balance;
+            mintWallet = coldWallet;           
+          }
+        }
+      }
+
+      setMintedMaxDegens(mintAmount);
+      setMintColdWallet(mintWallet);
+    }
+    })();
+  }, [coldwallets, whitelist])  
+
+  const addDegen = () => {
+    if(mintedDegens < mintedMaxDegens) {
+      setMintedDegens(clampMintAmount(mintedDegens + 1));
+    }
+  } 
 
-  const addDegen = () => setMintedDegens(clampMintAmount(mintedDegens + 1));
-  const subtractDegen = () => setMintedDegens(clampMintAmount(mintedDegens - 1));
+  const subtractDegen = () => {
+    if(mintedDegens > 0) {
+      setMintedDegens(clampMintAmount(mintedDegens - 1));
+    }
+  } 
 
   const onMintChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
     if (e.target.value.length > 4) {
       return;
     }
-
     const value = Number(e.target.value);
-    setMintedDegens(clampMintAmount(value));
+    if(value <= mintedMaxDegens ) {
+      setMintedDegens(clampMintAmount(value));
+    }
   }, []);
 
+  const onMint = async () => {
+    console.log("mint wallet:", mintColdWallet, mintedMaxDegens)
+    const ethersProvider = new ethers.providers.Web3Provider(provider);
+    const pfpContract = new ethers.Contract(pfpAddress, pfpAbi, ethersProvider);
+    
+    if(!mintedDegens || !mintColdWallet)
+      return;
+
+    setLoading(true)
+    const leafNodes = Object.keys(whitelist).map((addr: any) => keccak256(toUtf8Bytes(addr +'_'+ whitelist[addr])));
+    const merkleTree = new MerkleTree(leafNodes, keccak256, {sortPairs: true});
+    let claimingAllowance = whitelist[mintColdWallet as keyof typeof whitelist]
+    let claimingColdWallet = keccak256(toUtf8Bytes(mintColdWallet + '_' + claimingAllowance));
+    let hexProof = merkleTree.getHexProof(claimingColdWallet);
+    try {
+      let tx = await pfpContract.claimTokens(hexProof, mintedDegens, claimingAllowance)
+      let res = await tx.wait()
+      if (res.transactionHash) {
+        setLoading(false)
+        alert("mint success!")
+      }
+    } catch (err) {
+      console.log(err)
+      setLoading(false)
+    }
+  } 
+
```

Comment:
@blockchainFull2022 dont use any type please

@soulofmischief check this logic. Good bit of this is just foreign to me